### ISSUE TITLE
Improve name generation for local variables by giving types that implement Collection a plural name

### ISSUE THREAD

**Issue by natanfudge** [FabricMC/tiny-remapper#22]

**Comment by therealfarfetchd** [comment]
`Map<K, V>` → `kToV` / `kToVMap`

**Comment by Pyrofab** [comment]
`Iterator` → `it` maybe

**Comment by natanfudge** [comment]
`Iterator → it maybe`
That just seems less readable than `iterator`?

**Comment by therealfarfetchd** [comment]
I'd go for `iter`. Everyone seems to use that, and it's readable.

**Comment by Runemoro** [comment]
If a variable is contains the result of a method named `getXyz`, it should be named `xyz`.

**Comment by therealfarfetchd** [comment]
`ItemPlacementContext`, `ItemUsageContext`, `*Context` → `ctx`
`BlockPos` → `pos`
`BlockState` → `state`
`PlayerEntity` → `player`
`Direction` → `dir`
`BlockView` → `view`

**Comment by Pyrofab** [comment]
Building upon the PlayerEntity suggestion, `(.*)Entity` -> `$1`

**Comment by liach** [comment]
We should just allow the remapper to accept a local variable name suggestor

**Comment by natanfudge** [comment]
>ItemPlacementContext, ItemUsageContext, *Context → ctx
BlockPos → pos
BlockState → state
PlayerEntity → player
Direction → dir
BlockView → view

 - I think `context` is fine
- yes 
- yes
- no, `dir` is directory
- yes

**Comment by modmuss50** [comment]
I think expanding the auto name generator to include these in a yarn dist is the best way to go about this?

**Comment by natanfudge** [comment]
Gonna need lv remapping first

**Comment by modmuss50** [comment]
Ah yeah, didnt read the local bit. Could still do it for params possibly? might not be needed tho

**Comment by sfPlayer1** [comment]
Tiny remapper's name generation should remain independent of MC, so hardcoding rules for its typical classes is out - only important Java ones may receive some. The same applies to anything that can't be inferred without simple analysis, i.e. needing more than the descriptor+signature of the variable to be named.

You can still keep submitting ideas that are more extensive, we can implement more comprehensive schemes elsewhere in the tooling.

**Comment by therealfarfetchd** [comment]
>     * I think `context` is fine

I mean sure it's fine, but you want that name to not get in the way since most if not all of these are a container for other values whose names are the actually important part you want to look at when you're reading the code.

>     * no, `dir` is directory

Wouldn't be the first time an abbreviation could mean two different things based on context. And in most cases you're not going to be handling any directories in the same places as handling `Direction`s so they won't conflict.

`String` → `s` (not sure if this already happens)
`StringBuilder` → `sb`
`? extends Exception` → `e`

**Comment by Runemoro** [comment]
> Tiny remapper's name generation should remain independent of MC, so hardcoding rules for its typical classes is out

Can't the tiny format be extended to include a "default variable name" property for class mappings?

**Comment by natanfudge** [comment]
Why is there a need for a default variable name? Whoever creates the mappings file can:
- Put the name if it was inputted by a human
- Otherwise, if this name is special (like *context), put the desired name (`context` in this case)
- Otherwise, leave empty and let tiny-remapper handle it.

**Comment by natanfudge** [comment]
```
String → s (not sure if this already happens)
StringBuilder → sb
? extends Exception → e
```
I'm really against such short names in general...

**Comment by Runemoro** [comment]
@natanfudge I'm not sure what you're saying... therealfarfetchd suggested rules for automatically generating simpler names for certain classes, like `BlockPos` -> `pos`, and Player replied that they can't be hardcoded into tiny-remapper. I then suggested that they can be part of the mapping files.

**Comment by natanfudge** [comment]
I think what _you_ suggested, is to _extend_ the tiny format, with _new_ syntax. I replied that what Player wants is achievable using the existing format.

**Comment by Runemoro** [comment]
> is achievable using the existing format.

It isn't. With the existing format, we can only map things individually. Here we want a rule for *all* BlockPos variables to be named `pos` (plus a digit).

Unless you're suggesting moving the name generation to when yarn is built, but in that case this issue should be for yarn.

**Comment by natanfudge** [comment]
Is weave still alive?

**Comment by Runemoro** [comment]
No, weave has been merged into Enigma and is no longer necessary.

**Comment by natanfudge** [comment]
The original idea of this issue was about generic things (See the opening comment), I agree that minecraft-specific stuff can be moved to yarn.

**Comment by Runemoro** [comment]
It doesn't make sense to have variable name generation both in yarn and tiny-remapper...

**Comment by natanfudge** [comment]
Well, variable name generation is already partially implemented in tiny-remapper.

**Comment by Runemoro** [comment]
And it can be kept there. But we'll need a way to provide the custom rules to tiny-remapper (for example, by specifying the rules in the tiny files).

**Comment by sfPlayer1** [comment]
I like the idea of adding the default variable name, not sure how it'd interact with enigma / manual mapping authoring though. Some external tool with the knowledge+analysis capabilities of e.g. matcher would still have to handle complex cases like inferring names from getter returns etc - that would apply to field names as well.

**Comment by RedstoneParadox** [comment]
Gonna plus one not renaming int arrays to `is` as that would surely cause issue in Kotlin where `is` is a keyword.

**Comment by natanfudge** [comment]
This is about parameters/variables, which you can't interact with from a Kotlin mod.

**Comment by Runemoro** [comment]
Suggestions for short names:

 - `String` -> `s`
 - `Object` -> `o`
 - `Character`, `char`, result of `String.codePointAt` -> `c`
 - `Byte`, `byte` -> `b`
 - `Integer`, `int`, `Short`, `short`, `Long`, `long` -> `i`, `j`, `k`, `n`, `m`, `ii`, `jj`, ...
 - `Float`, `float`, `Double`, `double` -> `a`, `b`, `c`
 - `Function` -> `f`, `g`, `h`

**Comment by Runemoro** [comment]
> Map<K, V> → kToV / kToVMap

For maps, it should just be `Vs`, if there is only one map with value of type `V`. If there's several, it can be `VsByK`.

**Comment by Marcono1234** [comment]
Tiny Remapper currently increments the letters for primitive local variables (except for booleans), e.g. `(int ?, int ?)` would become `(int i, int j)`.
However, in my opinion this behavior is not so great because it often creates non-descriptive or irritating variables names if there are multiple variables, e.g.:  
`(char ?, char ?, double ?, double ?, float ?)` -> `(char c, char d, double e, double f, float g)`

It would probably better to just number them, e.g.:  
`(char ?, char ?, double ?, double ?, float ?)` -> `(char c, char c2, double d, double d2, float f)`

**Comment by sfPlayer1** [comment]
I found the distinct characters much more readable than incrementing the suffix after having experimented with both, could be made configurable though.

**Comment by coderbot16** [comment]
I think that after a certain number of local variables, numbering is more readable than incrementing letters. I think both approaches could be appropriate.
